boot/freeldr/freeldr/ntldr/wlmemory.c

```diff
@@ -238,6 +239,42 @@ WinLdrSetupMemoryLayout(IN OUT PLOADER_PARAMETER_BLOCK LoaderBlock)
         return FALSE;
     }
 
+    // Always map the ACPI table if present, otherwise Windows will crash.
+    if (Extension->AcpiTable)
+    {
+        PDESCRIPTION_HEADER AcpiTable = Extension->AcpiTable;
+        PRSDT_DESCRIPTOR Rsdt = Extension->AcpiTable;
+        PXSDT_DESCRIPTOR Xsdt = Extension->AcpiTable;
+
+        Status = MempSetupPaging((ULONG_PTR)Extension->AcpiTable >> MM_PAGE_SHIFT, 
+                                 MM_SIZE_TO_PAGES(Extension->AcpiTableSize),
+                                 FALSE);
+        if (!Status)
+        {
+            ERR("MempSetupPaging() failed\n");
+            return FALSE;
+        }
+
+        BOOLEAN is64BitsTable = _strnicmp(AcpiTable->Signature, "XSDT", 4) == 0;
+        ULONG Entries = (AcpiTable->Length - sizeof(DESCRIPTION_HEADER)) / (is64BitsTable ? 8 : 4);
+
+        for (ULONG i = 0; i < Entries; i++)
+        {
+            PDESCRIPTION_HEADER Entry = is64BitsTable ? 
+                    (PVOID)((ULONG_PTR)(Xsdt->PointerToOtherSDT[i])) :
+                    (PVOID)((ULONG_PTR)(Rsdt->PointerToOtherSDT[i]));
+
+            Status = MempSetupPaging((ULONG_PTR)Entry >> MM_PAGE_SHIFT, 
+                                     MM_SIZE_TO_PAGES(Entry->Length),
+                                     FALSE);
+            if (!Status)
+            {
+                ERR("MempSetupPaging() failed\n");
+                return FALSE;
+            }
+        }
+    }
```

Comment:
I'm not fully sure I understand the need for this manual mapping:
Wouldn't these be mapped already with the code below? (+ using PaToVa to convert the table pointer entries?)

Reply:
No it's not mapped with the code below.

Reply:
Yeah it's really strange. My suggestion would be to take a look at `AddMemoryDescriptor` and check if it works for you:
```
AddMemoryDescriptor(PcMemoryMap,
                    MAX_BIOS_DESCRIPTORS,
                    (PFN_NUMBER)(Entry / MM_PAGE_SIZE), // ACPI descripton header
                    (PFN_NUMBER)(Entry->Length / MM_PAGE_SIZE),
                    LoaderFirmwarePermanent);
```

Reply:
This works in 3 steps/phases:
1. Enumeration of memory regions from BIOS and hardcoded. `AddMemoryDescriptor` is used here to construct an initial memory map.
2. A "lookup table" is constructed, describing each page individually. This table is used subsequently to allocate pages (`MmAllocateMemoryWithType`, `MmAllocatePagesInLookupTable`)
3. Finalization: the final memory map is enumerated, pages that the kernel needs (see `MempSetupPagingForRegion` are mapped into kernel space) and a list of memory descriptors for the loader block is constructed.

To ensure that the memory descriptors for the kernel match what is mapped in the kernel (the kernel must not have any pages mapped, that the memory descriptors describe as free) the ACPI tables must be marked as `FirmwarePermanent` in the "lookup table" before step 3. No additional mapping is needed, as these pages will be mapped in step 3.